cupyx/scipy/stats/_stats_py.py

```diff
@@ -0,0 +1,147 @@
+
+    """
+
+    y = x[~cupy.isnan(x)]
+    if y.size == 0:
+        return cupy.array([True])
+    else:
+        return (y[0] == y).all(keepdims=True)
+
+
+def _quiet_nanmean(x):
+    """Compute nanmean for the 1d array x, but quietly return
+    nan if x is all nan. The return value is a 1d array with
+    length 1, so it can be used in cupy.apply_along_axis.
+
+    """
+
+    y = x[~cupy.isnan(x)]
+    if y.size == 0:
+        return cupy.array([cupy.nan])
+    else:
+        return cupy.mean(y, keepdims=True)
+
+
+def _quiet_nanstd(x, ddof=0):
+    """Compute nanstd for the 1d array x, but quietly return
+    nan if x is all nan. The return value is a 1d array
+    with length 1, so it can be used in cupy.apply_along_axis
+
+    """
+
+    y = x[~cupy.isnan(x)]
+    if y.size == 0:
+        return cupy.array([cupy.nan])
+    else:
+        return cupy.std(y, keepdims=True, ddof=ddof)
+
+
+def zmap(scores, compare, axis=0, ddof=0, nan_policy='propagate'):
+    """Calculate the relative z-scores.
+
+    Return an array of z-scores, i.e., scores that are standardized
+    to zero mean and unit variance, where mean and variance are
+    calculated from the comparison array.
+
+    Parameters
+    ----------
+    scores : array-like
+        The input for which z-scores are calculated
+    compare : array-like
+        The input from which the mean and standard deviation of
+        the normalization are taken; assumed to have the same
+        dimension as `scores`
+    axis : int or None, optional
+        Axis over which mean and variance of `compare` are calculated.
+        Default is 0. If None, compute over the whole array `scores`
+    ddof : int, optional
+        Degrees of freedom correction in the calculation of the
+        standard deviation. Default is 0
+    nan_policy : {'propagate', 'raise', 'omit'}, optional
+        Defines how to handle the occurrence of nans in `compare`.
+        'propagate' returns nan, 'raise' raises an exception, 'omit'
+        performs the calculations ignoring nan values. Default is
+        'propagate'. Note that when the value is 'omit', nans in `scores`
+        also propagate to the output, but they do not affect the z-scores
+        computed for the non-nan values
+
+    Returns
+    -------
+    zscore : array-like
+        Z-scores, in the same shape as `scores`
+
+    """
+
+    a = compare
+
+    if a.size == 0:
+        return cupy.empty(a.shape)
+
+    contains_nan, nan_policy = _contains_nan(a, nan_policy)
+
+    if contains_nan and nan_policy == 'omit':
+        if axis is None:
+            mn = _quiet_nanmean(a.ravel())
+            std = _quiet_nanstd(a.ravel(), ddof=ddof)
+            isconst = _isconst(a.ravel())
+        else:
+            mn = cupy.apply_along_axis(_quiet_nanmean, axis, a)
+            std = cupy.apply_along_axis(_quiet_nanstd, axis, a, ddof=ddof)
+            isconst = cupy.apply_along_axis(_isconst, axis, a)
```

Comment:
Is it possible to implement an equivalent thing with `cupy.apply_along_axis(_isconst, axis, a)` without using `apply_along_axis` as it is very inefficient?

Reply:
Great idea! I'll take a look at it.

Reply:
I guess this line would help you with modifying to omit nan.

https://github.com/cupy/cupy/blob/babc4994d14dd248ed94c84a921bf64613f853be/cupyx/scipy/stats/_stats_py.py#L139

Reply:
Ah, right! Thanks for sharing.